workspace/src/iot_shadow_service/iot_shadow_service/safe_cracker.py

```diff
@@ -0,0 +1,94 @@
+#!/usr/bin/env python3
+#
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: MIT-0
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy of this
+# software and associated documentation files (the "Software"), to deal in the Software
+# without restriction, including without limitation the rights to use, copy, modify,
+# merge, publish, distribute, sublicense, and/or sell copies of the Software, and to
+# permit persons to whom the Software is furnished to do so.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+# INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+# PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+# HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+# SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+#
+
+import json
+import rclpy
+from rclpy.node import Node
+from rclpy.qos import ReliabilityPolicy
+from threading import Lock
+from iot_shadow_service_msgs.msg import ShadowUpdateSnapshot
+from iot_shadow_service_msgs.srv import UpdateShadow
+
+DIGITS_ON_DIAL = 100
+MAX_SPEED = 3  # digits per spin
+SPIN_FREQUENCY = 0.1  # Hz
+
+
+class SafeCracker(Node):
+    def __init__(self):
+        super().__init__("safe_cracker")
+        self._target_digit = 0
+        self._current_digit = 0
+        self._lock = Lock()
+        self._clockwise = True
+
+        self.create_subscription(
+            ShadowUpdateSnapshot,
+            "shadow_update_snapshot",
+            self.on_shadow_update,
+            ReliabilityPolicy.BEST_EFFORT
+        )
+
+        self._update_client = self.create_client(UpdateShadow, "publish_to_shadow")
+        while not self._update_client.wait_for_service(timeout_sec=1.0):
+            self.get_logger().info("Update service not available, waiting again...")
+
+        self._update_timer = self.create_timer(SPIN_FREQUENCY, self.turn_towards_target)
+        self.get_logger().info("Starting to crack safe!")
+
+    def on_shadow_update(self, update: ShadowUpdateSnapshot) -> None:
+        desired = json.loads(update.desired)
+        self.get_logger().info(f"Desired shadow update is {desired}")
+        desired_digit = desired.get("digit", 0)
+        if desired_digit != self._target_digit:
+            with self._lock:
+                self._target_digit = desired_digit
+                self._clockwise = not self._clockwise
```

Comment:
So we flip the direction of the robot everytime there is a shadow update and desired is not the same as target?

Reply:
Flip the direction - yes. The target is the local copy of the desired digit - this if statement checks if there has been a change. It's useful to check in case a different shadow field has been updated; without the check, an update to another field would trigger a direction change even though the desired digit has not been updated